### ISSUE TITLE
Gather function in Lightning module with gradients

### ISSUE THREAD

**Issue by ananyahjha93** [PyTorchLightning/pytorch-lightning#4725]
## 🚀 Feature

Add a gather function with gradients in LightningModule which can be implemented with the help of:

https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/pl_bolts/models/self_supervised/simclr/simclr_module.py#L25

### Motivation

The default ```all_gather``` in pytorch breaks the gradient graph as it does not propagate any gradient backwards.

### Pitch

Allow users to use self.gather in LightningModule.

```
x = self.gather(x)
```
As suggested by @williamFalcon, this can be implemented like:

```
def gather(self):
    self.trainer.gather():


Trainer.gather(self, x):
    x = self.gather.accelerator(x)
```

All accelerators except CPU should implement this. 

The gradients can be synced with:

```
x = self.gather(x, sync_grads=True)
```

@edenlightning @tchaton @teddykoker

**Comment by teddykoker** [comment]
Here's the prototype I've been working with for DDP. @ananyahjha93 I feel the gather function should return a tuple, not a `cat`'d tensor so that the user can do what they please (`cat`, `stack`, etc.).
```python
import os

import torch
import torch.distributed as dist


class DiffAllGather(torch.autograd.Function):
    @staticmethod
    def forward(ctx, tensor):
        gathered = [torch.zeros_like(tensor) for _ in range(dist.get_world_size())]
        dist.all_gather(gathered, tensor)
        return tuple(gathered)

    @staticmethod
    def backward(ctx, *grad_outs):
        grad_outs = torch.stack(grad_outs)
        dist.all_reduce(grad_outs)
        return grad_outs[dist.get_rank()]

def all_gather(tensor):
    return DiffAllGather.apply(tensor)


def _test_all_gather(rank, world_size):
    os.environ["MASTER_ADDR"] = 'localhost'
    os.environ['MASTER_PORT'] = '8088'
    dist.init_process_group("gloo", rank=rank, world_size=world_size)

    x = torch.tensor(42., requires_grad=True)
    xs = torch.stack(all_gather(x))

    xs *= rank # multiply by rank
    xs.sum().backward()

    print(f"rank: {rank}, x:{x.grad}")

    # gradient should be equal to the rank for each process
    # so total gradient should be the sum of arange
    assert torch.allclose(
        x.grad,
        torch.arange(world_size).sum().float()
    )

def test_all_gather():
    world_size = 3
    torch.multiprocessing.spawn(_test_all_gather, args=(world_size,), nprocs=world_size)
```

**Comment by edenlightning** [comment]
@ananyahjha93 update?
